File: 204. Count Primes.py
class Solution(object):
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n < 3:
            return 0
        prime = [True] * n
        prime[0] = prime[1] = False
        for base in xrange(2, int((n - 1) ** 0.5) + 1):
            if prime[base]:
                prime[base ** 2::base] = [False] * len(prime[base ** 2::base])
        return sum(prime)

File: 204. Count Primes.java
public class Solution {
    public int countPrimes(int n) {
        // sieve of eratosthenes
        boolean[] composite = new boolean[n];
        int count = 0;
        for (int m = 2; m < n; m++) {
            if (!composite[m]) {
                count++;
                for (int k = 2; k*m < n; k++) {
                    composite[k*m] = true;
                }
            }
        }
        return count;
    }
}
